Added automating testing infrastructure

logiclrd opened the pull request:
I saw the existing PR #639 `Add a simple "test suite" of sorts` and noted that it has some drawbacks. Most notably, tests are all run in-process, which means if a test fails because the code crashed, then the entire test run aborts.

I decided to have half a go at making a bit more solid test suite.

This PR adds an automated testing infrastructure that operates off of a different build target. To build for testing, you run the `testbuild` target:
```
make clean testbuild
```
In this code, `main` is redirected to a test harness.

Tests are defined in methods with a specific signature and collected up in a file `index.c`. To add a new test, you add the test method, and then you add it to the list in `index.c`. Tests use a set of macros to perform assertions and return results, and can also return ad hoc output which is gathered and displayed independently at the end of the test.

```
testresult_t foobs()
{
  test_printf("Hello from test %s!", "foobs");
  ASSERT((1 == 1) == 1);
  RETURN_SKIP;
}
```

To run tests, you simply invoke `schismtracker` produced by `make testbuild`. You can also run a specific test by passing it on the command-line:
```
./schismtracker test_bshift_arithmetic
```

(This mechanism is used internally for running the full suite; a separate child process is spawned for each individual test.)

There is certainly some polishing that can be done to this:

* Display captured test output even when crashing?? Could possibly handle this by writing it to a file and having the harness check the file...
* Colour coding.
* Needs to be tested on architectures other than Linux and Win32.
* Other stuff I'm sure.

I also am not super familiar with autoconf, so I'm not sure if I did all the right things there. Empirically, though, it works on my end. :-)
```
(base) logiclrd@visor:/code/schismtracker$ ./schismtracker 
TEST: test_bshift_arithmetic ............................... PASS (0 ms)
TEST: test_bshift_right_shift_negative ..................... PASS (0 ms)
TEST: test_bshift_left_shift_overflow ...................... PASS (0 ms)
Results: 3 passed, 0 failed
(base) logiclrd@visor:/code/schismtracker$ 
```

logiclrd committed:
Added OS abstraction os_shell to encapsulate launching a child process by command-line, waiting for it and returning its exit code as an indivisible operation. Added implementations win32_shell and posix_shell. Updated configure.ac to test for posix_spawn and waitpid.
Added header automated-testing.h and implementation in automated-testing/harness.c, automated-testing/output.c and automated-testing/testresult.c. Updated main.c and macosx-sdlmain.m to call the ENTRYPOINT determined by whether SCHISM_TEST_BUILD is defined.
Added target testbuild to Makefile.am which defines SCHISM_TEST_BUILD and then triggers an otherwise normal full build.
Added some test cases to kick things off in automated-testing/cases/bshift.c. Added file automated-testing/index.c that acts as a directory of all defined test cases. Added include/test-assertions.h.
Added all new files to the lists in Makefile.am.

mrpapersonic commented on automated-testing/testresult.c:
```diff
@@ -0,0 +1,36 @@
+	case SCHISM_TESTRESULT_NOT_RUN: return "NOT RUN";
+	case SCHISM_TESTRESULT_PASS: return "PASS";
+	case SCHISM_TESTRESULT_FAIL: return "FAIL";
+	case SCHISM_TESTRESULT_INCONCLUSIVE: return "INC.";
```
Can this say "INCLUSIVE" instead of incorporated?

logiclrd replied:
It can :-) I'll do that in a bit. The only thing is that it means that a column of text containing these values is suddenly potentially much longer.

mrpapersonic replied:
as long as it can fit in 120 columns or so, it's fine.

mrpapersonic commented on include/osdefs.h:
```diff
@@ -84,6 +84,12 @@ A return value of 0 indicates that the event should NOT be processed by the main
 #endif
 
+#if defined(SCHISM_WIN32)
+# define os_shell win32_shell
```
Isn't this virtually equivalent to
```c
char *cwd = dmoz_get_current_directory();
os_run_hook(cwd, name, arg);
free(cwd);
```
? It would be nice if it could just be reduced to that for all platforms :)

logiclrd replied:
It isn't, because that doesn't wait on the child process and return its exit code.

mrpapersonic replied:
Maybe we could separate out the common gory bits of `os_run_hook` and `os_shell` and put them into a separate function, maybe `os_exec`, that does all the dirty work, and then the former two functions can just be implemented as stubs that call `os_exec`. (this hypothetical function should probably also store the program return value in a passed pointer, because I think on some OSes it isn't limited to just 0..255 (?) like it is on linux)

logiclrd replied:
Huh, I didn't realize that POSIX limited it to 8 bits. It's `int` in all the APIs. But sure enough, Linux only gives you 8 bits. The POSIX API is limited thusly, because it packs a bunch of stuff into the `status` parameter to `waitpid` and only gives 8 bits to the exit code. I just checked FreeBSD, though, and if you use its `wait6` API function, which is considerably more capable than `waitpid`, you can pass in a structure that receives the full 32-bit exit code. Windows as well uses `DWORD` values for process exit codes (thread exit codes too). OS/2 takes a `ULONG` to `DosExit` which is in turn returned via `DosWaitChild`. Linux does have an API call `waitid` that _could_ return the full value, but I tested it and it does not. So, it seems Linux is the wimpy one here 😁 

(An AI summary told me that POSIX's 2016 update actually specifies that `waitid` _not_ limit the exit status to 8 bits, but that Linux is dragging its heels on actually implementing it.)

The code I wrote here actually already breaks on Linux because I just assumed I had an `int` to work with and encode some test statuses to values larger than 255. 😮 

As for making a generalized abstraction, it should probably just return an `int`. That might not be 32-bit on all platforms, but it should be _more than enough_ on any platform. 🙂 If we use `waitid` on Linux, then it might in the future magically start returning the full 32 bits.

logiclrd replied:
Looking at the implementations, it looks like `run_hook` is a straight superset of `shell`, and could be implemented in terms of calling it instead of doing its own machinations.

logiclrd replied:
Incidentally, `MAX_PATH` is not the maximum length of the path on Windows. It is the maximum length of a _path component_. This is _widely widely_ misunderstood, but in the Win32 model, a path is a null-terminated `TSTR[32767]`.

logiclrd replied:
Okay, I've reworked the Win32 implementation. I made the following changes:

* It just straight-up always uses the wide character API. In this day and age, I don't see a compelling reason to support the ANSI character API, and in extreme cases, an ANSI build may encounter files and paths it can't access because the characters can't be mapped into the current code page. Unicode will just work.
* This means that the nasty undebuggable macros for the variants aren't needed. Instead, there's a pithy little function `win32_create_process_wait`, and `win32_shell` and `win32_run_hook` both use it to actually invoke the process.
* The process is created using `CreateProcessW`, which supports passing in the initial CWD directly. No more `chdir`-and-back nonsense. :-)

Now, on the other hand, the Linux implementations are _not_ so easily-collapsed.

* `posix_shell` uses the function `posix_spawn`, which simplifies the process of running a child process. Under the hood, I'm sure it's implemented as `fork`/`exec`, but it's much simpler to use at the top end.
* `posix_run_hook` can't use this function, though, because POSIX _doesn't_ provide anything like `CreateProcessW`'s `lpCurrentDirectory` parameter. Instead, the paradigm is to `chdir`, which makes sense, because the paradigm is also `fork`/`exec`, so you're your own process anyway, with your own CWD.

So, `posix_shell` is tight and compact because it leverages `posix_shell`, and `posix_run_hook` _can't_ leverage `posix_shell`. I think they just have to stand as they are.

logiclrd replied:
> It isn't, because that doesn't wait on the child process and return its exit code.

When I wrote this, incidentally, I had the strange idea in my head that `os_run_hook` was fire-and-forget. I don't know where that came from.

`os_run_hook` is pretty much a superset of `os_shell`, but the preceding comment explains why the POSIX versions are independent. On Windows, while you could technically write one in terms of the other, it is simpler, with fewer dynamic buffer allocations (I think), to use a common implementation function per 884b306.

logiclrd replied:
Hmm, also, this:

https://github.com/schismtracker/schismtracker/blob/6306e7a337077ae428d5b7188834ee5da5338d00/sys/posix/osdefs.c#L76

...isn't threadsafe. That is likely completely irrelevant with current uses; they won't ever overlap. But as a generalized function, this is going to have issues if two threads run it concurrently, since `wait` will return on _any_ child process exiting.

logiclrd replied:
I just committed a change that should address this. However, it is untested, because I'm not at my Linux workstation right now. It's a pretty simple change, I expect it to work, but I'll give it a test once I'm on `visor` again. 26d7f03

mrpapersonic replied:
> It just straight-up always uses the wide character API. In this day and age, I don't see a compelling reason to support the ANSI character API, and in extreme cases, an ANSI build may encounter files and paths it can't access because the characters can't be mapped into the current code page. Unicode will just work.

The Unicode APIs don't work on Win9x, but I'm 99% sure I'm the only one who even cares about that (or even routinely tests it). So I'll be the one to fix up Win9x when it breaks.

For the `posix_spawn` stuff, I mostly suggested that it be something like

```c
/* launches a subprocess and waits for it */
int posix_exec(/* possibly NULL */ const char *dir, const char *exe, /* possibly NULL */ const char *maybe_arg, int *status)
{
    int st;

    switch (fork()) {
    /* mostly just cruft from posix_run_hook with additional checks for NULL */
    }

    while (wait(&st) == -1);

    if (WIFEXITED(st) && status)
        *status = WEXITSTATUS(st);

    return 1;
}
```

where then `posix_run_hook` can be
```c
int posix_run_hook(const char *dir, const char *exe, const char *maybe_arg)
{
    int st;
    char *s;

    if (asprintf(&s, "./%s", exe) < 0)
        return 0;

    if (!posix_exec(dir, s, maybe_arg, &st)) {
        free(s);
        return 0;
    }

    free(s);

    return (st == 0);
}
```

and posix_shell:
```c
int posix_shell(const char *exe, const char *arg)
{
    int st;

    if (!posix_exec(NULL, exe, arg, &st))
        return -1; /* ? */

    return st;
}
```

unless im just totally missing something. Then we could (hopefully) remove the platform-specific `os_run_hook` entirely and just base it all off of `os_exec`.

logiclrd replied:
Mm, okay. I'll shuffle stuff around. :-)

logiclrd replied:
Oh, but. The Win32 implementation has all the batch file-specific logic in it. So, `os_exec` will have to be able to handle that contextually...

mrpapersonic replied:
It can just do what the `posix_run_hook` implementation i put above does, just putting `.bat` on the end instead of a `./` at the front :)

logiclrd replied:
Right, but if we're going to have a consistent `platform_exec` interface, then I'm thinking we don't actually need `platform_run_hook` and `platform_shell` at all. They can be generically implemented. Commit coming up in a bit.

mrpapersonic commented on Makefile.am:
```diff
@@ -506,6 +508,11 @@ endif
 
 ## aaaaaaaaahhhhhhhhhhhhhhhhhhh!!!!!!!1
 schismtracker_SOURCES = \
+	automated-testing/harness.c        \
```
hm, this shouldn't be apart of the normal build. It should have an `--enable-tests` that explicitly enables it.

logiclrd replied:
I can see arguments both ways. It was intentional that it was included, because that means that it _must build_, even when it isn't in use. If this ends up being part of PR hooks, then that isn't an issue because test builds will be done all the time anyway :-P

logiclrd replied:
Maybe there's a best-of-both-worlds option. What if it always built the code, but it left the objects out of the linking phase for non-test builds? Come to think of it, in full-optimizing mode, wouldn't it _effectively_ do that anyway? If you define a symbol, but nothing references it, wouldn't it get stripped out during linking??

logiclrd replied:
I just gave it a quick test, and it seems it does not omit unused objects automatically, but the build could probably be configured so that it builds everything but then just doesn't link in the testing-related objects in the default target.

logiclrd replied:
This seemingly-simple idea appears to be rather tricky to fit into the paradigm of autoconf. :-P

logiclrd replied:
Okay this looks like it ought to do it.

I've defined a new target in Makefile.am called `schismtrackertest` (on Windows, `schismtrackertest.exe`). A default build builds both `schismtracker` and `schismtrackertest`. The `schismtrackertest` source files contain all of the test harness & test case files that we don't want built into `schismtracker`. They do _not_ contain the rest of the code; instead, the object files built by the `schismtracker` build are pulled into the link of `schismtrackertest` directly using `schismtrackertest_LDADD`.

With this change, all files are built only once, except for one file, `automated-testing/entrypoint.c`, which defines a thunk to call either `schism_main` or `schism_test_main`, depending on the build type. When `schismtrackertest_LDADD` is defined to include `schismtracker_OBJECTS`, this particular object file is explicitly filtered out (otherwise there'd be two definitions of the thunk when linking).

There's only one thing I haven't figured out that I am aware of: As far as I can tell, `make install` is going to install `schismtrackertest` in the `bin` directory alongside `schismtracker`. I spend a good amount of time trying to figure out how to exclude it from `install` while still taking advantage of the build pipeline, but I didn't find a way. That's a problem for another day. :-)

```
$ autoreconf && ./configure
...
$ make
...
$ ./schismtrackertest
TEST: test_bshift_arithmetic ..................................... PASS (0 ms)
TEST: test_bshift_right_shift_negative ........................... PASS (0 ms)
TEST: test_bshift_left_shift_overflow ............................ PASS (0 ms)
Results: 3 passed, 0 failed
$ ./schismtracker
𝄞 yey music ♯ ♫
```

With these changes, every `make` builds everything. If a change causes a test to stop compiling, you find out about it right away. But the testing infrastructure and test cases do not clutter up the binary.

logiclrd committed:
Expanded the string representation of SCHISM_TESTRESULT_INCONCLUSIVE to the full string "INONCLUSIVE" in testresult_str.
Added macro TESTRESULT_STR_MAX_LEN to automated-testing.h and updated run_test in harness.c to calculate the length of the line of dots to right-align the testresult_str column, allowing for its maximum potential width.

logiclrd committed:
Added a new target schismtrackertest(.exe) to Makefile.am.
Moved automated-testing files from schismtracker_SOURCES to schismtrackertest_SOURCES in Makefile.am.
Added new file automated-testing/entrypoint.c that defines a thunk that calls either schism-main or schism-test-main. This one file, then, gets compiled with different output in both regular and testing builds. Updated Makefile.am to include automated-testing/entrypoint.c in both schismtracker_SOURCES and schismtrackertest_SOURCES, and to exclude its object from the schismtracker side when collecting objects for schismtrackertest_LDADD.

logiclrd committed:
Oops, I fixed this but must have accidentally done it to the generated `Makefile` rather than the source.

mrpapersonic commented on Makefile.am:
```diff
@@ -672,3 +675,24 @@ schismtracker_LDADD = $(LIBM) $(libs_sdl3) \
+schismtrackertest_CPPFLAGS = $(schismtracker_CPPFLAGS)
+schismtrackertest_OBJCFLAGS = $(schimtracker_OBJCFLAGS)
+schismtrackertest_DEPENDENCIES = $(schismtracker_DEPENDENCIES)
+schismtrackertest_LDADD = $(filter-out automated-testing/schismtracker-entrypoint.o,$(schismtracker_OBJECTS)) $(schismtracker_LDADD)
```
Isn't this a GNU make extension? (though i see we're also using it above...??????!!!)

anyway, we shouldn't really be using GNU make extensions since the point of automake is to be able to create portable makefiles that work everywhere (tho most everyone's just using gnu tools anyway now, so it isn't really that big of a deal.)

logiclrd replied:
It looks like that's the case. What's the alternative?

mrpapersonic replied:
no idea, I think i'll just put it off until an issue actually pops up with it :^)

logiclrd replied:
Okeydoke :-)

logiclrd committed:
Added the main schismtracker binary as a dependency of schismtrackertest, since schismtrackertest isn't directly depending on the source code files and won't be rebuilt otherwise.

logiclrd committed:
Rewrote win32_shell and win32_run_hook:
- They always use Unicode in the underlying call.
- The underlying call is provided by a common worker function win32_create_process_wait.

logiclrd committed:
Updated result_to_exit_code and exit_code_to_result to use exit code values that fit into 8 bits. Needed just for Linux support. :-)

logiclrd committed:
Switched posix_run_hook from wait() to waitpid(), so that as a general-use function it is now threadsafe.

logiclrd committed:
Replaced platform_shell and platform_run_hook with a platform_exec interface that encapsulates the needed functionality for both.
Added non-platform implementations of os_shell and os_run_hook based on platform_exec functions.
Reworked posix_exec to use waitid instead of waitpid, since there is some possibility this may do away with the 8-bit restriction on exit codes in the future. Changed the function check in configure.ac for waitpid to waitid.
Updated run_test_child in automated-testing/harness.c to test for an os_exec macro rather than an os_shell macro to determine if a call to os_shell is going to be implemented.
Updated the in-process testing warning in run_test_child to flush stderr immediately.

logiclrd committed:
Added missing #include to posix/osdefs.c.

logiclrd committed:
Oof. posix/osdefs.c is being compiled with WIN32 builds too. Moved #include inside #if SCHISM_WIN32.

logiclrd committed:
Fixed the #if defined() checks for mapping os_exec to posix_exec in osdefs.h.
Updated the guard around the definition of posix_exec in posix/osdefs.c to match.

logiclrd committed:
Fixed detection of whether os_exec is available in sys/osdefs.c. Updated osdefs.h to #define HAVE_OS_EXEC.

logiclrd committed:
Added schismtrackertest to .gitignore.

logiclrd committed:
Updated osdefs.c to provide some sort of filler for ENOTSUP on OS/2, which seems to lack this in its POSIX layer.

mrpapersonic committed:
tests: refactor a bit, win32 is completely untested

I've changed the function parameters for os_exec to be a variable
argument function. This makes it look a lot more like the POSIX
execl. This also means we can take any number of arguments we want,
and (for the most part) all of the cruft can be kept inside of
the exec implementation. However, run_hook still needs to be
implemented separately.

os_shell is completely gone. There was no way to tell if the actual
program failed or the process creation failed. That ambiguity sucks,
especially when we want to fall back if the process creation failed.

Test output now uses a dynamic memory buffer instead of the linked
list shenanigans it was doing before. This is probably faster anyway.

I don't know if relying on argv[0] being the program name everywhere
is ok, but it's fine for now. It's okay on all the platforms that
matter. Plus, since we can fall back, who cares.

mrpapersonic committed:
tests: CRLF -> LF

it seems the ASIO stuff is also CRLF. stupid notepad++

mrpapersonic committed:
osdefs: whoops, forgot to define os_run_hook as a stub

mrpapersonic committed:
win32: fix(?) build

mrpapersonic committed:
win32: oops, again

mrpapersonic committed:
automake: add new headers to noinst_HEADERS

eh

mrpapersonic committed:
*: automated-testing -> test, make building test binary optional

package maintainers absolutely don't need the test binary, nor do
users building from the source tarball. we still need to automatically
run the test suite for github actions though.

mrpapersonic committed:
macosx-sdlmain: automated-testing -> test

mrpapersonic committed:
test: move some things around

also, since not every platform uses .o as an object extension
(notable suspects are OpenWatcom, which SHOULD use .obj, but for
some reason doesn't under our CI, and MSVC), I've changed the
.o in automake to be OBJEXT, which should work everywhere :')
